### PR TITLE
feat(docker-build): Orchestrated containers for DB, API and UI

### PR DESCRIPTION
**Pull request by abhi-markan**
# Orchestration

## Introduction
Due to the advent of *Keystone* (Headless CMS) being used as an `API` container which relies on a `SQL` database.
Since we have been spinning Alpine linux image for Keystone, few compatibility issues have emerged with the installation and spooling of `mariadb` and `mysql` on Alpine linux.

Due to the above issue docker containers were unable to boot and thus the consumption of API via `:5001` was unsuccessful.
Following remedial actions have been in order to triage this issue with few to-do to be addressed post this PR merge.

## Resolution
* Docker - YML schema version updated to `3.3`.
* Docker - Introduction of a `MySQL` container.
* Docker - API container port to changed to `5001`.
* Docker - API container keystone initialisation command.
* API `init.sh` starts keystone.
* Exceptional handling, error print.
* Apollo path update, removed `/graphql` already present in the environment variable.
* Environment variable name change from `DB_URI` to `DATABASE_URL`.
* Updated `.dockerignore`
* Updated `.gitignore`
* Updated `.env.sample`.
* MySQL `exip` database initialisation during first `database` container spin up.
* Introduction of `database` directory with SQL dump. (Only executed if `.database` volume does not exists) 
* Addition of `API` microservice build environment variable for `Keystone` build.

## To-Do
* MySQL secure installation
* MySQL configuration

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
 .vscode
 .DS_Store
 .env
-node_modules
+.api
+.database
```
Comment (ttbarnes) [inline .gitignore:5]: Do we need `.api` and `.database` here? I don't believe these get generated

File: src/api/.dockerignore [hunk 2]
```diff
@@ -1,7 +1,9 @@
 node_modules
+README.md
 Dockerfile
-*.md
+.git/
+docs/
+dist/
```
Comment (ttbarnes) [inline src/api/.dockerignore:6]: Same as above I don't think these get generated?

File: src/ui/.gitignore [hunk 3]
```diff
@@ -1,6 +1,6 @@
 .vscode
 .DS_Store
 .env
-node_modules
-
-generated_reports
+.api
+.database
+node_modules
```
Comment (ttbarnes) [inline src/ui/.gitignore:6]: Same as above, seems that `.git`, `docs/` and `dist/` don't get generated

Comment (abhi-markan) [reply src/ui/.gitignore:4]: Good spot, have removed.

File: src/api/.dockerignore (hunk 2 above)
Comment (abhi-markan) [reply src/api/.dockerignore]: Good spot, have removed.

File: .gitignore (hunk 1 above)
Comment (abhi-markan) [reply .gitignore:5]: We do need both of the above, to avoid being tracked.
Above both are both Docker persistent storage volumes.

**Review by ttbarnes — APPROVED** [review]
LGTM great work! 🚀

**Review by ttbarnes — COMMENTED** [review]
Build is failing - also need to run `npm run lint:fix` in the API (sorry that's my fault from the WIP `EMS-548` branch. Sorry didn't reliase until after approved

File: docker-compose.yml [hunk 4]
```diff
@@ -1,31 +1,42 @@
...
+      - ./.api/src/.keystone:/app/src/.keystone:ro
     environment:
-      - MYSQL_ROOT_PASSWORD=example
-      - PORT=5005
-      - DB_URI=mysql://root@localhost:5005/exip
-      - SESSION_SECRET=asdfaslfdjjasjoiwefjwoerij23j8j2oi2j
-      - GOV_NOTIFY_API_KEY
-    entrypoint: npx nodemon src/keystone.ts -L
+      PORT: 5001
+      DATABASE_URL: mysql://root:password@database:3306/exip
```
Comment (IgnatG) [inline docker-compose.yml:32]: can we split this into `DATABASE_URL` and `DATABASE_NAME`? the name may change. the URL shouldn't change though

Comment (abhi-markan) [reply docker-compose.yml:32]: Good spot, have removed and added them to `.env` variables only.

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,7 @@
 .vscode
 .DS_Store
 .env
+.api
+.database
 node_modules
+e2e-tests/cypress/*
```
Comment (Zainzzkk) [inline .gitignore:7]: will this not ignore e2e tests?

File: docker-compose.yml [hunk 6]
```diff
@@ -1,31 +1,45 @@
...
-      - MYSQL_ROOT_PASSWORD=example
-      - PORT=5005
-      - DB_URI=mysql://root@localhost:5005/exip
-      - SESSION_SECRET=asdfaslfdjjasjoiwefjwoerij23j8j2oi2j
-      - GOV_NOTIFY_API_KEY
-    entrypoint: npx nodemon src/keystone.ts -L
+      PORT: 5001
+      DATABASE_URL:
+      GOV_NOTIFY_API_KEY:
+
```
Comment (Zainzzkk) [inline docker-compose.yml:36]: we also need to add COMPANIES_HOUSE_API_URL and COMPANIES_HOUSE_API_KEY

Comment (abhi-markan) [reply docker-compose.yml:36]: Added.

File: .gitignore (hunk 5 above)
Comment (abhi-markan) [reply .gitignore:7]: Good spot, have added exceptional directories.

**Review by Zainzzkk — APPROVED** [review]
LGTM

**Review by ttbarnes — APPROVED** [review]
LGTM great work @abhi-markan ! 🚀 🎉 👏

**Review by ttbarnes — APPROVED** [review]
Fantastic work! LGTM

**Review by IgnatG — COMMENTED** [review]
Looks good overall. I left some comments

File: .github/workflows/pr_tests.yml [hunk 7]
```diff
@@ -84,21 +84,35 @@ jobs:
...
+          SESSION_SECRET: ${{ secrets.SESSION_SECRET }}
+          BASIC_AUTH_KEY: ${{ secrets.BASIC_AUTH_KEY }}
+          BASIC_AUTH_SECRET: ${{ secrets.BASIC_AUTH_SECRET }}
+          MULESOFT_API_CIS_URL: ${{ secrets.MULESOFT_API_CIS_URL }}
+          MULESOFT_API_CIS_KEY: ${{ secrets.MULESOFT_API_CIS_KEY }}
+          MULESOFT_API_CIS_SECRET: ${{ secrets.MULESOFT_API_CIS_SECRET }}
+          MULESOFT_API_MDM_EA_URL: ${{ secrets.MULESOFT_API_MDM_EA_URL }}
+          MULESOFT_API_MDM_EA_KEY: ${{ secrets.MULESOFT_API_MDM_EA_KEY }}
+          MULESOFT_API_MDM_EA_SECRET: ${{ secrets.MULESOFT_API_MDM_EA_SECRET }}
+          GOOGLE_ANALYTICS_ID: ${{ secrets.GOOGLE_ANALYTICS_ID }}
```
Comment (IgnatG) [inline .github/workflows/pr_tests.yml:109]: do we need this here?

File: src/api/Dockerfile [hunk 8]
```diff
@@ -3,7 +3,9 @@ FROM node:18.9-alpine3.16
 
 # Variables
 ARG GITHUB_SHA
-ENV GITHUB_SHA=$GITHUB_SHA
+ARG GOV_NOTIFY_API_KEY
+ENV GITHUB_SHA $GITHUB_SHA
+ENV GOV_NOTIFY_API_KEY $GOV_NOTIFY_API_KEY
```
Comment (IgnatG) [inline src/api/Dockerfile:8]: we don't have these ENV as part of Dockerfile in DTFS. do we need them here? or can they be used as part of docker-compose.yml file?

Comment (abhi-markan) [reply src/api/Dockerfile:8]: We do need them here as these are build environment variables and docker compose are only passed during runtime.

File: .github/workflows/pr_tests.yml (hunk 7 above)
Comment (abhi-markan) [reply .github/workflows/pr_tests.yml:109]: Yes these needs to be passed for E2E containers, without which containers will not spin up properly.

File: src/api/Dockerfile [hunk 9]
```diff
@@ -3,7 +3,9 @@ FROM node:18.9-alpine3.16
 
 # Variables
 ARG GITHUB_SHA
-ENV GITHUB_SHA=$GITHUB_SHA
+ARG GOV_NOTIFY_API_KEY
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply src/api/Dockerfile:6]: Yes for build.

File: src/ui/server/graphql/apollo.ts [hunk 10]
```diff
@@ -10,7 +10,7 @@ dotenv.config();
 const urlRoot = process.env.API_URL;
 
 const createHttpLinkOptions = {
-  uri: `${urlRoot}/graphql`,
+  uri: `${urlRoot}`,
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply src/ui/server/graphql/apollo.ts:13]: Nope, being supplied in env variable.

File: src/api/Dockerfile (hunk 8 above)
Comment (IgnatG) [reply src/api/Dockerfile:8]: Can we pass the values in the docker-compose.yml file? like this?
```yml
    environment:
      PORT: 5001
      DATABASE_URL: ${DATABASE_URL}
      GOV_NOTIFY_API_KEY: ${GOV_NOTIFY_API_KEY}
      COMPANIES_HOUSE_API_URL: $GOV_NOTIFY_API_KEY}
      COMPANIES_HOUSE_API_KEY: ${COMPANIES_HOUSE_API_KEY}
```

this ^ should pick up the values from .env file and pass it to the Docker image.

File: .github/workflows/pr_tests.yml (hunk 7 above)
Comment (IgnatG) [reply .github/workflows/pr_tests.yml:109]: hmm. this is not good. we shouldn't need the GA ID to be able to spin them up. if it doesn't exist, then the containers shouldn't crash. but maybe something to address in a different PR

**Pull request merged**